runtime-light/stdlib/curl/curl-multi-functions.h

```diff
@@ -4,14 +4,287 @@
 
 #pragma once
 
+#include <chrono>
 #include <cstdint>
 #include <functional>
 #include <optional>
 
 #include "runtime-common/core/runtime-core.h"
+#include "runtime-light/coroutine/task.h"
+#include "runtime-light/stdlib/curl/curl-context.h"
+#include "runtime-light/stdlib/curl/curl-state.h"
+#include "runtime-light/stdlib/curl/defs.h"
+#include "runtime-light/stdlib/curl/details/diagnostics.h"
 #include "runtime-light/stdlib/diagnostics/logs.h"
+#include "runtime-light/stdlib/fork/fork-functions.h"
+#include "runtime-light/stdlib/web-transfer-lib/defs.h"
+#include "runtime-light/stdlib/web-transfer-lib/details/web-property.h"
+#include "runtime-light/stdlib/web-transfer-lib/web-composite-transfer.h"
+#include "runtime-light/stdlib/web-transfer-lib/web-simple-transfer.h"
 
-inline Optional<array<int64_t>> f$curl_multi_info_read([[maybe_unused]] int64_t /*unused*/,
-                                                       [[maybe_unused]] Optional<std::optional<std::reference_wrapper<int64_t>>> /*unused*/ = {}) {
-  kphp::log::error("call to unsupported function : curl_multi_info_read");
+inline auto f$curl_multi_init() noexcept -> kphp::coro::task<kphp::web::curl::multi_type> {
+  auto open_res{co_await kphp::forks::id_managed(kphp::web::composite_transfer_open(kphp::web::transfer_backend::CURL))};
+  if (!open_res.has_value()) [[unlikely]] {
+    kphp::web::curl::print_error("could not initialize a new curl multi handle", std::move(open_res.error()));
+    co_return 0;
+  }
+  const auto descriptor{(*open_res).descriptor};
+  CurlInstanceState::get().multi_ctx.get_or_init(descriptor);
+  co_return descriptor;
+}
+
+inline auto f$curl_multi_add_handle(kphp::web::curl::multi_type multi_id, kphp::web::curl::easy_type easy_id) noexcept -> kphp::coro::task<Optional<int64_t>> {
+  auto& curl_state{CurlInstanceState::get()};
+  if (!curl_state.multi_ctx.has(multi_id)) [[unlikely]] {
+    co_return false;
+  }
+  auto& multi_ctx{curl_state.multi_ctx.get_or_init(multi_id)};
```

Comment:
Let’s move this declaration below to not call `curl_state.multi_ctx.get_or_init` if `curl_state.easy_ctx.has(easy_id))` is false